captureoutgoing.go

```diff
@@ -0,0 +1,186 @@
+type Transport struct {
+	Transport   http.RoundTripper
+	LogRequest  func(req *http.Request)
+	LogResponse func(resp *http.Response)
+}
+
+// The default logging transport that wraps http.DefaultTransport.
+var DefaultTransport = &Transport{
+	Transport: http.DefaultTransport,
+}
+
+type contextKey struct {
+	name string
+}
+
+var ContextKeyRequestStart = &contextKey{"RequestStart"}
+
+// RoundTrip is the core part of this module and implements http.RoundTripper.
+func (t *Transport) RoundTrip(request *http.Request) (*http.Response, error) {
+	ctx := context.WithValue(request.Context(), ContextKeyRequestStart, time.Now())
+	request = request.WithContext(ctx)
+
+	// Outgoing Request Time
+	outgoingReqTime := time.Now().UTC()
+
+	response, err := t.transport().RoundTrip(request)
+	if err != nil {
+		return response, err
+	}
+
+	// Outgoing Response Time
+	outgoingRspTime := time.Now().UTC()
+
+	// Skip capture outgoing event
+	shouldSkipOutgoing := false
+	if _, found := moesifOption["Should_Skip_Outgoing"]; found {
+		shouldSkipOutgoing = moesifOption["Should_Skip_Outgoing"].(func(*http.Request, *http.Response) bool)(request, response)
+	}
+
+	// Skip / Send event to moesif
+	if shouldSkipOutgoing {
+		if debug{
+			log.Printf("Skip sending the outgoing event to Moesif")
+		}
+	} else {
+		
+		// Check if the event is to Moesif
+		if !(strings.Contains(request.URL.String(), "moesif.net")) {
+
+			if debug {
+				log.Printf("Sending the outgoing event to Moesif")
+			}
+
+			// Get Request Body
+			var outgoingReqBody interface{}
+			var reqEncoding string
+			outgoingReqBody = nil
+			if logBodyOutgoing && request.Body != nil {
+				copyBody, err := request.GetBody()
+				if err != nil {
+					if debug{
+						log.Printf("Error while getting the outgoing request body: %s.\n", err.Error())
+					}
+				}
+			
+				// Read the request body
+				readReqBody, reqBodyErr := ioutil.ReadAll(copyBody)
+				if reqBodyErr != nil {	
+					if debug {
+						log.Printf("Error while reading outgoing request body: %s.\n", reqBodyErr.Error())
+					}
+				}
+			
+				// Parse the request Body
+				reqEncoding = "json"
+				if jsonReqParseErr := json.Unmarshal(readReqBody, &outgoingReqBody); jsonReqParseErr != nil {
+					if debug {
+						log.Printf("About to parse outgoing request body as base64 ")
+					}
+					outgoingReqBody = b64.StdEncoding.EncodeToString(readReqBody)
+					reqEncoding = "base64"
+					if debug {
+						log.Printf("Parsed outgoing request body as base64 - %s", outgoingReqBody)
+					}
+				}
+			
+				// Return io.ReadCloser while making sure a Close() is available for request body
+				request.Body = ioutil.NopCloser(bytes.NewBuffer(readReqBody))
+			
+			}
+  
+			// Get Response Body
+			var outgoingRespBody interface{}
+			var respEncoding string
+			outgoingRespBody = nil
+			if logBodyOutgoing && response.Body != nil {
+				// Read the response body
+				readRespBody, err := ioutil.ReadAll(response.Body)
+				if err != nil {
```

Comment:
may be for later..parsing request/response body is common code, and be kept in a fcn, makes sdk smaller?